### PR TITLE
Add/gutenboarding basic analytics

### PR DESCRIPTION
**Pull request by roo2**
*Depends on https://github.com/Automattic/wp-calypso/pull/39392 being merged* I'll then need to rebase this onto `master`

This is a continuation of the calypso-analytics PR https://github.com/Automattic/wp-calypso/pull/39001
It adds examples of analytics to gutenboarding and:

* Pull the code that calculates the `last_pageview_path_with_count`  and `this_pageview_path_with_count` metadata into the calypso-analytics package.
* Add types to the calypso-analytics package
* Initialize analytics in gutenboarding
* Also added types for `User` and `UserData` ( ended up not using them in this PR but they are useful for editor autocompletion etc)
* Add example recordTracksPageView and recordTracksEvent calls to gutenboarding.

The next step will be to go through and add all of the analytics that we need in gutenboarding following the same pattern.

## Testing

- [x] Test existing page view events in calypso are setting `last_pageview_path_with_count` and `this_pageview_path_with_count` correctly
- [x] Test IE11 still attaches `?utm_` parameters (make sure there's no conflict with #39481)
- [x] Smoke test existing analytics in calypso
- [x] Test that `/gutenboarding` fires a pageView event and that `gutenboarding/signup` fires a calypso_gutenboarding_signup_start event with the pageview params set
Note: analytics in gutenboarding will not include *user* metadata such as number of sites created. This is because we will not have a *user* until an account is created

### CODE REVIEW

**Comment by sirreal** [comment]
I've added the "in progress" label, as I assumed this to be in progress 🙂
Using a draft PR will prevent code owners automatic review requests from being sent.

File: packages/calypso-analytics/tsconfig.json
Comment (sirreal) [inline packages/calypso-analytics/tsconfig.json]: Please use `dist/types` like the rest of the packages:

```suggestion
		"declarationDir": "dist/types",
```

File: packages/calypso-analytics/package.json
Comment (sirreal) [inline packages/calypso-analytics/package.json]:
```suggestion
	"types": "dist/types",
```

File: packages/calypso-analytics/package.json
Comment (sirreal) [inline packages/calypso-analytics/package.json]:
```suggestion
```

File: packages/calypso-analytics/tsconfig-cjs.json
Comment (sirreal) [inline packages/calypso-analytics/tsconfig-cjs.json]:
```suggestion
		"tsBuildInfoFile": "../../.tsc-cache/calypso-analytics-cjs"
```

File: packages/calypso-analytics/tsconfig.json
Comment (sirreal) [inline packages/calypso-analytics/tsconfig.json]:
```suggestion
		"tsBuildInfoFile": "../../.tsc-cache/calypso-analytics"
```

File: packages/load-script/package.json
Comment (sirreal) [inline packages/load-script/package.json]: This package needs `tslib` added to its `dependencies`.

File: packages/calypso-analytics/global.d.ts
Comment (sirreal) [inline packages/calypso-analytics/global.d.ts]: Is this intended to be a general purpose library? `BUILD_TIMESTAMP` global is very Calypso-specific, isn't it?

File: packages/calypso-analytics/global.d.ts
Comment (sirreal) [inline packages/calypso-analytics/global.d.ts]: I believe it would be preferable to move this declaration to the module that adds `_tkq` to the window.

File: packages/calypso-analytics/src/tracks.ts
Comment (sirreal) [inline packages/calypso-analytics/src/tracks.ts]: This would be a good place to add `_tkq` to the window type.

File: packages/calypso-analytics/package.json
Comment (sirreal) [inline packages/calypso-analytics/package.json]: This isn't accurate. At the very least `page-view-params` adds an event listener, `tracks` adds `_tkq` to the window object. I haven't audited the rest of the modules, but this should be fixed.

File: packages/calypso-analytics/src/utils/current-user.ts
Comment (sirreal) [inline packages/calypso-analytics/src/utils/current-user.ts]: Will this work? Do we want it to work? Generally, I wouldn't expect packages to import anything from the client app.

**Review by sirreal — COMMENTED** [review]
Thanks for working on this. I've left a lot of comments and suggestions.

This seems to contain a lot of unrelated changes which makes review a bit more difficult. Was that intended?

File: packages/calypso-analytics/global.d.ts
Comment (roo2) [reply packages/calypso-analytics/global.d.ts]: It seems like _tkq could be initialized in a few different places, each place where it is used, the code will initialize it if it's not defined with: `window._tkq = window._tkq || [];` It looks like calypso-analytics *is* the primary place it's initialized because the other places are all apps.

File: packages/calypso-analytics/global.d.ts
Comment (roo2) [reply packages/calypso-analytics/global.d.ts]: hmm yes, a more general approach would be to get the caller to pass in `BUILD_TIMESTAMP`... but that's not very nice for useability, I think it is better if we do just try to read the `window.BUILD_TIMESTAMP` if it's there, just to keep the function easy to use at the expense of leaking calypso concepts into the analytics library. What do you think?

File: packages/calypso-analytics/src/utils/current-user.ts
Comment (roo2) [reply packages/calypso-analytics/src/utils/current-user.ts]: yea that was a bit crazy I took out that reference.

**Comment by roo2** [comment]
Thanks for the reviews @sirreal! I've fixed most of the feedback and asked a followup question about the other. I also fixed the base so that the unrelated changes aren't included.

**Review by sirreal — COMMENTED** [review]
Another round of drive-by review, I haven't been testing or looking closely at the code.

**Comment by sirreal** [comment]
I've just noticed that this is "in progress" so perhaps my reviews were premature 😅 Hope they're helpful!

File: packages/calypso-analytics/src/tracks.ts
Comment (roo2) [reply packages/calypso-analytics/src/tracks.ts]: The fromEntries property isn't in the Typescript definitions for Object so typescript complains, it was only added recently https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries. (I tested this does work in IE11 though, we must be transpiling it to something else)

File: client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx
Comment (roo2) [reply client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx]: nice, I'll look into that

File: packages/load-script/package.json
Comment (roo2) [reply packages/load-script/package.json]: ahhh I see

File: client/lib/user/user.d.ts
Comment (roo2) [reply client/lib/user/user.d.ts]: Ok, you're right

> I've just noticed that this is "in progress" so perhaps my reviews were premature 😅 Hope they're helpful!

No, reviews are good :D sorry I forgot to update the label!

**Comment by roo2** [comment]
> I've just noticed that this is "in progress" so perhaps my reviews were premature 😅 Hope they're helpful!

No, reviews are good :D sorry I forgot to update the label!

File: packages/calypso-analytics/src/tracks.ts
Comment (roo2) [reply packages/calypso-analytics/src/tracks.ts]: Nice typescript foo 👍 
hmm this has exposed quite a few places where we weren't checking for `window` to be defined.. I wonder under what circumstances `window` would be undefined...? Some of this code is tested via `client/lib/analytics/test/index.js` and that test code also assumes that `window` exists. I can't think of any reason `window` would be undefined 😂.

File: packages/calypso-analytics/src/page-view-params.ts
Comment (roo2) [reply packages/calypso-analytics/src/page-view-params.ts]: haha yep now the git blame will lead back to me 😂 

"but it's not the cleanest code - sorry" 
~ Eru Penkman (Approved by @sirreal)

I basically just tested in the browser that it produces the same behavior as it previously did and moved it all into it's own file.

File: client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx
Comment (roo2) [reply client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx]: I tried to get this working but useLocation().pathname returns only the pathname within the current router. So here it returns '/about' when I want '/gutenboarding/about'. I looked for ways to get the '/gutenboarding' component of the path (the `BrowserRouter.basename` defined in `client/landing/gutenboarding/index.tsx:47`) but I couldn't find anything suitable beyond using `window.location.pathname` I think at this stage we should just hardcode it here. The main goal of the PR is to prove the new tracks API with examples. In the future we might do something slightly different, e.g. record every gutenboarding `calypso_page_view` from within `gutenboarding/index.js`.

File: client/landing/gutenboarding/index.tsx
Comment (roo2) [reply client/landing/gutenboarding/index.tsx]: phew I think that is a different `User` :joy: I was including the `analytics` user returned by `client/lib/user/user.js`, but I wonder if the `client/lib/user/user.js` would also work 🤔. 

Yea I'm with you on getting more type info in and hopefully this `@automattic/calypso-types` is the right place for those unused `client/lib/user/user.js` type definitions I wrote.

I wrote a note to come back to this :)

File: packages/calypso-analytics/package.json
Comment (roo2) [reply packages/calypso-analytics/package.json]: hmm ok so `sideEffects` is a flag controlling whether webpack can safely prune unused exported functions or not. If it's `true` than all functions will always be included in the bundle because webpack can't work out if they are necessary just for their "sideEffects". https://stackoverflow.com/questions/49160752/what-does-webpack-4-expect-from-a-package-with-sideeffects-false

Our only side effect is initializing `window._tkq` when the library is loaded; `client/lib/analytics/index.js` does in fact depend on this having happened.

I've actually gone and contained all usage of `window._tkq` inside a new `pushEventToTracksQueue()` function, and updated the dependent code in `client/lib/analytics/` to use this. 

So now the code still has the "Side Effect" of adding to the `window` object, but it doesn't have `sideEffects` because there are no implicit dependencies between functions :blush:.

There are other places in `apps/` where we use `window._tkq` but fortunately these places check to ensure that it's been initialized. I probably should update them to all use `pushEventToTracksQueue` but I don't want to introduce the dependency on `@automattic/calypso-analytics` to those apps right now 😋

File: packages/calypso-analytics/tsconfig.json
Comment (sirreal) [inline packages/calypso-analytics/tsconfig.json]: Regarding https://github.com/Automattic/wp-calypso/pull/39451/files#r381582967 - let's try this:

```suggestion
		"lib": [ "DOM", "ESNEXT" ],
```

[`Object.fromEntries` is available](https://unpkg.com/browse/typescript@3.7.5/lib/lib.es2019.object.d.ts), it's a question of finding the right configuration.

**Comment by ramonjd** [comment]
Sorry, this PR is too big for me to give a meaningful review today. Thanks to @sirreal for picking through the code, and for the informative reviews!

But I did go through Gutenboarding - 

`calypso_page_view` fires for `gutenboarding/about` and `calypso_gutenboarding_signup_start` fires correctly 👍 

Also smoke testing analytics throughout Calypso to look for regressions. All page load and click tracking events are firing, and I can't see any obvious issues.

File: client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx
Comment (p-jackson) [inline client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx]: Just a quick comment for now: I wonder if this could be done in a general way rather than having to call `recordTracksPageViewWithPageParams` on mount of each component. Plus it's susceptible to breaking when we make changes to the route.

I was thinking about listening to route updates instead of component mounts e.g. adding something like this to the `<Gutenboard>` component:

```typescript
import { useParams, useHistory } from 'react-router-dom';

const { location: { pathname } } = useHistory();
useEffect( () => {
	recordTracksPageViewWithPageParams( `/gutenboarding${pathname}` );
}, [ pathname ] );
```

Adding it to `gutenboard.tsx` instead of `index.tsx` means it will only be triggered by routes that we actually support i.e. no `/gutenboarding/abcxyz` track events.

Another thing it will do differently is send the locale slug as part of the event too. I'm not sure whether that's how `recordTracksPageViewWithPageParams` is supposed to be used, but I assume so.

What do you think?

Comment (p-jackson) [reply client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx]: Just checked the old signup framework, and yes it's supposed to include the language slug.

Comment (roo2) [reply client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx]: nice, that looks good, I'll add that code and test it, thanks.

File: client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx
Comment (roo2) [reply client/landing/gutenboarding/onboarding-block/vertical-select/index.tsx]: update on this issue, I've moved page view tracking into the `Gutenboarding` component so each `gutenboarding/*` page view will be tracked automatically (thanks Philip)

File: packages/calypso-analytics/src/tracks.ts
Comment (roo2) [reply packages/calypso-analytics/src/tracks.ts]: this is fixed now, using your `ESNEXT` change in tsconfig

File: packages/calypso-analytics/package.json
Comment (p-jackson) [reply packages/calypso-analytics/package.json]: @roo2 marked as resolved but this still doesn't look right to me. Unless I've misunderstood something, we can't mark this package as `sideEffects=false` because importing code from this package does indeed have side-effects (e.g. the event listener).
Does this need to be `sideEffects=true`?

File: packages/calypso-analytics/src/tracks.ts
Comment (p-jackson) [reply packages/calypso-analytics/src/tracks.ts]: Some tests run in a `jsdom` environment and some don't, so just because all tests currently assume `window` is available, doesn't mean there won't be tests in the future that don't have `jsdom` but would still like to import this package. Adding `jsdom` to tests does slow them down a little.

File: packages/calypso-analytics/src/tracks.ts
Comment (p-jackson) [inline packages/calypso-analytics/src/tracks.ts]: If `window` isn't defined maybe this should go down the same code path as when `window.crypto` isn't defined.

File: packages/calypso-analytics/src/utils/hash-pii.ts
Comment (p-jackson) [inline packages/calypso-analytics/src/utils/hash-pii.ts]: Now that this is TS we can delete the `{...}` in the jsdoc

File: packages/calypso-analytics/src/utils/current-user.ts
Comment (p-jackson) [inline packages/calypso-analytics/src/utils/current-user.ts]: We don't need the {...} in the jsdoc here

File: packages/calypso-analytics/src/utils/current-user.ts
Comment (p-jackson) [inline packages/calypso-analytics/src/utils/current-user.ts]: Are there any cases where `hashedPii` can be missing? If we make it non-optional it would catch any cases where we forget to set.

File: packages/calypso-analytics/src/utils/current-user.ts
Comment (p-jackson) [inline packages/calypso-analytics/src/utils/current-user.ts]: I was wondering if it'd be better to just define a wee interface for this function's parameter type. Something like
```typescript
interface SetCurrentUserParam {
	ID: string;
	username: string;
	email: string;
}
```

It doesn't need to be exported.

File: packages/calypso-analytics/src/utils/current-user.ts
Comment (p-jackson) [inline packages/calypso-analytics/src/utils/current-user.ts]: This logic changed very subtly from what it used to be in JS.

```suggestion
		isNaN( parseInt( currentUser.ID, 10 ) ) ||
```

Without knowing much about the format of the `ID` string, I'd say it's safer to stick with `parseInt` so we can use the explicit radix param.

File: packages/calypso-analytics/src/page-view-params.ts
Comment (p-jackson) [reply packages/calypso-analytics/src/page-view-params.ts]: I see now, this comment was just copy-pasted from `lib/analytics`. Not sure what the nasty coupling is that it's refering to.

`analytics.pageView.record` and `analytics.pageLoading.record` are both in in `lib/analytics`, so maybe we should just leave this code there? Can we just reimplement the parts that are needed for `recordTracksPageViewWithPageParams` to work as expected?

File: client/landing/gutenboarding/gutenboard.tsx
Comment (p-jackson) [inline client/landing/gutenboarding/gutenboard.tsx]: Just noticed that when I navigate to `/gutenboarding` and look at the `calypso_page_view` event in the network inspector, it's sending a trailing slash in the path i.e. `path=/gutenboarding/`. Not sure how much it matters (I don't know every way tracks data is used 😅), it might be annoying for folks designing their funnels using tracks data. But fwiw events from the `/start` onboarding flow don't have a trailing slash.

**Review by p-jackson — COMMENTED** [review]
Phew that's a lot of analytics 😅 Breaking the PR up into:

1. changes related to `*_with_count` params
2. converting to TypeScript
3. record events in gutenboarding

would make the reviewing a bit easier. Particularly because there might be a lot more discussion around (1) and how we actually want analytics to work in the new package world (which is a good thing to talk about), but that discussion doesn't need to hold up (3).

File: packages/calypso-analytics/src/tracks.ts
Comment (roo2) [reply packages/calypso-analytics/src/tracks.ts]: ahh I can't do that because in the case where `window.crypto` is undefined, it still uses the `window.btoa` function on line `:62`

File: client/landing/gutenboarding/gutenboard.tsx
Comment (roo2) [reply client/landing/gutenboarding/gutenboard.tsx]: hmm it's different but I don't think we should add code to change it unless it does turn out to cause any issues for the analytics people, the other routes like `/gutenboarding/start` don't have a trailing `/`, so the `pathname` must be `/` here to give us `/gutenboarding/`. I think it's alright as long as they can track the unique URLs

File: packages/calypso-analytics/src/utils/current-user.ts
Comment (roo2) [reply packages/calypso-analytics/src/utils/current-user.ts]: Good catch!

**Comment by roo2** [comment]
> Phew that's a lot of analytics 😅 Breaking the PR up into:
> 
>     1. changes related to `*_with_count` params
> 
>     2. converting to TypeScript
> 
>     3. record events in gutenboarding
> 
> 
> would make the reviewing a bit easier. Particularly because there might be a lot more discussion around (1) and how we actually want analytics to work in the new package world (which is a good thing to talk about), but that discussion doesn't need to hold up (3).

Yes, I should have done the typescript conversion seperately to make this easier to review, I think that *3)*, recording the `calypso_page_view` event in gutenboarding is dependent on *1)*, pulling the page view params into `calypso-analytics`, I haven't made changes to that functionality works (apart from that integer conversion which you found, oops) but just pulled it all into one place to be reused. So yea I don't know if we want to change the way the `*_with_count` stuff works, now would be a good opportunity to change it but for now it's just been moved.

So it would have been good but I'd rather not pull the PR into separate parts now that it's mostly reviewed 😅

**Comment by roo2** [comment]
hmm yea, I signed up for a new site with the free domain `erutestredirectsite.design.blog` and then set it to my default site. when I set the route to `/`, I got this page:

Screen Shot 2020-02-25 at 4 13 18 PM

File: packages/calypso-analytics/src/tracks.ts
Comment (p-jackson) [reply packages/calypso-analytics/src/tracks.ts]: lol so it does, missed that 😂

File: packages/calypso-analytics/src/tracks.ts
Comment (roo2) [reply packages/calypso-analytics/src/tracks.ts]: ahh ok that makes sense, yea this code ideally shouldn't depend on jsdom for testing but it looks like it does for now. I'll just leave those guards around window.

File: packages/calypso-analytics/src/page-view-params.ts
Comment (roo2) [reply packages/calypso-analytics/src/page-view-params.ts]: Those comments seem to be incorrect/out of date. There is no `pageLoading.record` function, it isn't even in the git history, I checked out master and couldn't find any other references to the`mostRecentUrlPath` or `pathCounter` variables (appart from `analytics.timing.record`).

There is a coupling to `analytics.timing.record` because `analytics.timing.record` does re-use `mostRecentUrlPath` to submit its events. I think that's not ideal but it does it via the getMostRecentUrlPath function so at least it is explicit.

I updated the comments to reflect how the variables are being used.

Comment (p-jackson) [reply packages/calypso-analytics/src/page-view-params.ts]: Thanks. The safest thing is probably to leave it as it is for now. If there's someone that "owns" `analytics.timing.record` then we could ask what the GA timings are being used for and if there's a cleaner way we could implement it. Maybe someone in @Automattic/team-calypso knows who's using the GA timing measurements? Seems like they're related to a config flag called `perfmon`.

Lets not do any of that in this PR though.

**Comment by p-jackson** [comment]
Been manually testing the changes in Calypso and the `calypso_page_view` and GA timing events are working as expected (i.e. the same way they work on `master`).

Testing GA was a little tricky, but I did it by adding another `debug` line to `analytics.timing.record()` and enabling `perfmon` in `development.json`.

I think the last remaining thing in this PR is to do with `sideEffects` in the `package.json`. I think it's changed between `false` and `true` a couple times during the history of the PR so I'm struggling to figure out the conversation that's already happened 😅
@roo2 is the plan to switch it to `sideEffects=true`

File: packages/calypso-analytics/package.json
Comment (roo2) [reply packages/calypso-analytics/package.json]: Ahh sorry I accidentally hit mark as resolve instead of posting my comment here 🤐. 
It looks like `sideEffects` is used to determine if webpack is allowed to do tree shaking and not load files that aren't needed. If the module has `sideEffects` it means that there are implicit dependencies between functions. So "side effects" like attaching the event handler to window don't count as `sideEffects`, as long as other functions don't depend on it and it could safely be left out by the tree shaking build. 
https://stackoverflow.com/questions/49160752/what-does-webpack-4-expect-from-a-package-with-sideeffects-false

We used to have a cross function dependency on `window._tkq` being initialized at the top of the file, but I've refactored it so that `window._tkq` is not used directly, rather, the wrapping function `pushEventToTracksQueue` is used. I've also updated other places in calypso (`client/lib/analytics/index.js`) to use `pushEventToTracksQueue`. This way there are no implicit dependencies (despite still using the global) and the tree shaker can trace down to include only the functions that it needs.

So yes I believe it is correct to set `sideEffects:false` now, despite the code having "side effects".

File: packages/calypso-analytics/src/page-view-params.ts
Comment (roo2) [reply packages/calypso-analytics/src/page-view-params.ts]: ahh do you mean to leave the PR as it is now or to leave this page_view params code in `lib/analytics/index.js`? 

It looks like @gravityrail wrote these comments about tracking `last_pageview_path_with_count` and `this_pageview_path_with_count` originally in 2016. I wonder if he remembers?

**Comment by roo2** [comment]
Thanks @p-jackson for the reviews and especially for manually testing `analytics.timing.record`!

I think that we should have `sideEffects=false` because that flag is used to track whether tree shaking can be used on the library, even if there are side effects like the one you pointed out  https://github.com/Automattic/wp-calypso/pull/39451#discussion_r384204666

I really struggled to find the conversation too and ended up writing another comment, I installed this user script via tampermonkey to add an "Expand, Expand" button to github to show all conversations 😂 https://github.com/findepi/user.js/blob/master/github-expand-expand-expand.user.js

File: packages/calypso-analytics/package.json
Comment (p-jackson) [reply packages/calypso-analytics/package.json]: > "side effects" like attaching the event handler to window don't count as sideEffects, as long as other functions don't depend on it and it could safely be left out by the tree shaking build

Hmmm I hadn't thought of it like that 🤔

I guess in theory it shouldn't make a difference whether there's a latent window listener there or not. But I'm worried it's the sort of thing that'd be really hard to track down if it _did_ cause a difference between dev and prod builds. I'm thinking of cases where it was tricky fixing a Renovate PR because the modules getting updated had side effects.

It's not what I'd expect from something that's labelled side-effect free, but I see from the SO post that there's a fuzzy line since things like `console.log`s are generally ok. Do you have any opinions on where to draw the line @jon?

Another way would be to add the event listener in `initializeAnalytics()`, but that'd only work if it was being called early enough.

Comment (roo2) [reply packages/calypso-analytics/package.json]: hmm It looks like setting `sideEffects=true` might be a safer and clearer option and I think that this module is so small and all in use that tree shaking wouldn't really make a difference, I'll just set it to true I think.

Comment (p-jackson) [reply packages/calypso-analytics/package.json]: Sweet, you did bring up a good point though. I'm interested in figuring out where we should be drawing that line in general.

**Review by p-jackson — APPROVED** [review]
This LGTM

The failing tests look like they have something to do with Customer Home/Stats changes, which I thought would have been fixed by #39653. Either way, I've certainly been able to navigate to the stats page in my testing, so the failures look unrelated.

File: packages/calypso-analytics/package.json
Comment (sirreal) [reply packages/calypso-analytics/package.json]: You can also set it to a sideEffects to an array of files that have sideEffects, but just `true` is probably enough here. We aren't likely to make big gains here.

**Comment by p-jackson** [comment]
> "Expand, Expand"

Nice! Added.

**Pull request merged**
